Add support to checkpoint/restore containers

adrianreber opened the pull request:
This is still work in progress and just published for initial feedback.

runc uses CRIU to support checkpoint and restore of containers. This
brings an initial checkpoint/restore implementation to podman.

None of the additional runc flags are yet supported and container
migration optimization (pre-copy/post-copy) is also left for the future.

The current status is that it is possible to checkpoint and restore a
container. I am testing on RHEL-7.x and as the combination of RHEL-7 and
CRIU has seccomp troubles I have to create the container without
seccomp.

With the following steps I am able to checkpoint and restore a
container:
```
 # cat unconf.json
 {}
 # podman run --security-opt="seccomp=unconf.json" -d registry.fedoraproject.org/f26/httpd
 # podman checkpoint <container>
 # podman restore <container>
```
I am using CRIU, runc and conmon from git. Especially conmon needs
additional changes for checkpoint/restore which are still in a PR.

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber commented:
This is my first attempt to provide checkpoint/restore to podman. This needs conmon changes which are available here: https://github.com/kubernetes-incubator/cri-o/pull/1427

Right now I am mainly hoping for some feedback if my approach is correct or if this should be done differently.

Container checkpointing and restoring works but the network does not yet work after the restore as I am not setting up the network namespace during restore.

rhatdan commented:
@adrianreber Added [WIP} to heading to prevent accidentally getting merged.

mheon commented on libpod/container.go:
```diff
@@ -89,6 +89,16 @@ func (ns LinuxNS) String() string {
+	// CreateContainer creates a new container
+	CreateContainer CreateMode = iota
+	// RestoreContainer restores an existing container
+	RestoreContainer CreateMode = iota
```
Do we see there being more of these in the future? I can't think of obvious ones, so I would lean towards just leaving this a bool argument in OCIRuntime.createContainer

mheon commented on libpod/container_api.go:
```diff
@@ -678,3 +679,101 @@ func (c *Container) Sync() error {
+		//}
+	}
+
+	if c.state.State == ContainerStateRunning {
```
Probably worth including Paused in there too, I doubt we want to restore over a paused container

mheon commented:
Fundamental approach looks right - it looks like our other API functions, which is a good thing

I'm wondering if we want to add a build tag for this, to avoid having another hard dependency. Not sure yet.

adrianreber commented:
@mheon Thanks for the feedback, I will change the two things you mentioned and try to get networking running after restore and update this PR.

TomSweeneyRedHat commented on libpod/container_api.go:
```diff
@@ -678,3 +679,101 @@ func (c *Container) Sync() error {
+	// This also happens in Cechkpoint()
+	// Not sure where is the right place for this.
+	// If the container is stopped via runc, it needs to be here for a successful restart.
+	logrus.Debugf("Cleaning up %s", filepath.Join(c.bundlePath(), "attach"))
```
nit, think you want a space before attach here and ctl on 765

rhatdan commented:
@adrianreber Is this still something you are interested in?  Needs a rebase.

adrianreber commented:
@rhatdan I am still on it. I rebased it locally and I am currently trying to get the network in the same state as is was during the checkpoint.

adrianreber commented:
This will take some more time as runc and CRIU do not handle pre-created namespaces correctly like podman does. See: https://github.com/opencontainers/runc/issues/1786

Once I have the necessary changes on CRIU and runc I will continue here.

adrianreber commented:
I have finally updated this PR to introduce checkpoint and restore in podman.

All changes in required dependencies (conmon, runc, criu) have been committed to the respective repositories. Right now I am using all of those dependencies from git checkouts.

For CRIU a one-line patch is still under discussion: https://lists.openvz.org/pipermail/criu/2018-September/042613.html

I am also removing the [WIP] flag as for the initial implementation everything is ready.

I can checkpoint and restore containers and after restore I can still (again) connect to the network service provided by the container, using the same IP address.

I tried to address the review comments from the initial commit from March. Next steps from my point of view:

 * more advanced network setups
 * offer more of the CRIU functionality exposed through runc also in podman (i.e. migrating established TCP connections)
 * handle container migration between hosts. Right now I am only checkpointing and restoring on one system
 * add support for pre-copy and post-copy container migration

adrianreber commented:
Ah, I did not run the code linter. Will do that and update the PR.

TomSweeneyRedHat commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+	}
+
+	// This needs to be deleted so that the restore works correctly.
+	// This also happens in Chechkpoint()
```
nit Chechkpoint to Checkpoint

adrianreber replied:
Fixed in upcoming update. Thanks.

TomSweeneyRedHat commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+	os.Remove(filepath.Join(c.bundlePath(), "stats-dump"))
+	os.Remove(filepath.Join(c.bundlePath(), "stats-restore"))
+	// Network status file
+	os.Remove(filepath.Join(c.bundlePath(), "network.status"))
```
I concur with the no error out, but am just wondering if throwing a debug message on error might be helpful if things ever go wrong.  I think it's probably fine as is, but thought I'd raise the question for consideration.

adrianreber replied:
Sounds like a good idea.

rhatdan commented:
Would we also want this for a pod?

rhatdan commented on libpod/container_internal.go:
```diff
@@ -522,7 +522,7 @@ func (c *Container) init(ctx context.Context) error {
 
 	// With the spec complete, do an OCI create
-	if err := c.runtime.ociRuntime.createContainer(c, c.config.CgroupParent); err != nil {
+	if err := c.runtime.ociRuntime.createOCIContainer(c, c.config.CgroupParent, false); err != nil {
```
You don't need the other functions in createContainer?

adrianreber replied:
Not sure, this changed between my first PR and now (after rebase). I will have a closer look what is the right function.

adrianreber replied:
Using createContainer() also just works. It seems, however, nobody else is using the createContainer() function. Is that a function for external usage?

mheon replied:
I believe this is a rootless thing - if you have root, createContainer just wraps createOCIContainer, but for rootless it also does some additional work to ensure things are set up properly

rhatdan commented:
Command completions.
Tests
Man pages
Perhaps update tutorial.
Nice addition...

adrianreber commented:
@rhatdan : I will add all those things

mheon commented on libpod/oci.go:
```diff
@@ -766,3 +770,15 @@ func (r *OCIRuntime) execStopContainer(ctr *Container, timeout uint) error {
+// checkpointContainer checkpoints the given container
+func (r *OCIRuntime) checkpointContainer(ctr *Container) error {
+	// imagePath is used by CRIU to store the actual checkpoint files
+	imagePath := filepath.Join(ctr.bundlePath(), "checkpoint")
```
If we're going to use this more than once, maybe a helper function for getting it? `ctr.getCheckpointPath()` or similar?

adrianreber replied:
will be in next PR update

mheon commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+		}
+	}
+
+	if c.state.State == ContainerStateRunning {
```
I think we should also error out here on pause - I don't think restoring a paused container is valid?

adrianreber replied:
will be in next PR update

mheon commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+	// This needs to be deleted so that the restore works correctly.
+	// This also happens in Checkpoint()
+	// Not sure where is the right place for this.
+	// If the container is stopped via runc, it needs to be here for a successful restart.
```
I think what you're looking for is `removeConmonFiles()`

adrianreber replied:
Good to know it exists. Thanks. Will use it.

mheon commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+	c.state.State = ContainerStateStopped
+
+	// This needs to be deleted so that the restore works correctly
+	os.Remove(filepath.Join(c.bundlePath(), "attach"))
```
I don't think we need this here - having it on restore (it's already in restart) is enough

adrianreber replied:
Makes sense. Will remove it.

mheon commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+	}
+
+	if keep {
+		return c.save()
```
I think this needs to be unconditional to make sure we pick up the ContainerStateStopped above

adrianreber replied:
This is just my strange code flow. It happens again later. Will change it to make it more readable.

mheon commented on libpod/container_api.go:
```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
+}
+
+// Restore restores a container
+func (c *Container) Restore(keep bool) (err error) {
```
Is there an easy way to check that a checkout exists at the start of the function, so we can error out if one doesn't? Should we track checkpoints in the database?

adrianreber replied:
I can test for the checkpoint directory and for one of the central checkpoint files. If both exist it should be a real checkpoint. Not sure if it should be tracked in the database.

mheon replied:
Testing for existence sounds sufficient. We might want to track them if we ever start adding more than one, but checking for existence is plenty good for now.

mheon commented on cmd/podman/main.go:
```diff
@@ -62,6 +62,7 @@ func main() {
 
 	app.Commands = []cli.Command{
 		attachCommand,
+		checkpointCommand,
```
I think these are missing?

adrianreber replied:
Yes, this is still wrong. Need to fix it. Thanks for all the reviews.

mheon commented:
bot, add author to whitelist

adrianreber commented:
Updated PR and force pushed, I will add commits for man pages, tests, completion and tutorial tomorrow.

adrianreber commented:
I have a question about a travis failure:

```libpod/container_api.go:968:55:warning: unused variable or constant invalid operation: c.state (variable of type *containerState) has no field or method NetNS (varcheck)```

This is an error on a non-Linux system. Only in libpod/container_linux.go the state gets the additional field NetNS, on non-Linux systems this is missing. As CRIU is a Linux only thing, do I need to move my Checkpoint and Restore methods to container_linux.go? Instead of container_api.go? Or do I need to create container_api_linux.go and have empty stubs in container_api_unsupported.go for Checkpoint() and Restore()

adrianreber commented:
In travis the integration tests are failing and it is not clear to me why exactly they are failing. Is this something temporary or related to my changes?

rhatdan commented:
Travis has been acting up all day.

adrianreber commented:
So everything besides completion should be ready. Let's see if travis likes my test cases.

Completion should be ready soon. I will update the PR.

adrianreber commented:
And now with completions. Let me know if anything else is missing.

adrianreber commented:
I moved the checkpoint and restore commands to be sub-commands of 'podman container' and the last travis run on my branch finally succeeded. Let's see if the travis result is repeatable.

mheon commented:
@adrianreber Our CI is a bit of a mess right now, so I would not be surprised if various tests are exploding, and most of it is our fault, not yours

adrianreber commented:
@mheon thanks for the information. I was a bit worried when all tests except travis failed at the same time.

adrianreber commented:
Finally all tests are green. Is anything missing to get this merged?

mheon commented on cmd/podman/checkpoint.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+func checkpointCmd(c *cli.Context) error {
+	if os.Geteuid() != 0 {
```
I think you might need a rootless.IsRootless() check here too

mheon commented on cmd/podman/checkpoint.go:
```diff
@@ -0,0 +1,73 @@
+			lastError = errors.Wrapf(err, "error looking up container %q", arg)
+			continue
+		}
+		logrus.Debugf("Trying to checkpoint container %s", ctr)
```
Move the debug log here into ctr.Checkpoint() - if we ever call it from outside of podman, we'd still like the debug logs

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	}
+
+	// Copy /etc/resolv.conf to the container's rundir
+	_, err = c.generateResolvConf()
```
Can we use makeBindsMounts() here instead of generateResolvConf() and generateHosts()?

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+		if IP != nil {
+			env := fmt.Sprintf("IP=%s", IP)
+			// Tell CNI which IP address we want.
+			os.Setenv("CNI_ARGS", env)
```
We're adding support to request static IPs from CNI (have a PR open against ocicni, but not merged yet) - can you add a TODO comment here to convert this code to use the new static IP handling once it merges?

adrianreber replied:
I added a comment referencing that ocicni pull request.

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	return c.save()
+}
+
+func (c *Container) restore(keep bool) error {
```
To make the defers that check error work, can you change the return to (err error) instead of just error?

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	c.state.State = ContainerStateRunning
+
+	if !keep {
+		// Delete all checkpoint related files. At this point, in theory, all files
```
So doing a restore() consumes the checkpoint and renders it unusable? Should this be documented in the manpages?

mheon commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+
+func (c *Container) restore(keep bool) error {
+
+	if c.state.State == ContainerStateRunning || c.state.State == ContainerStatePaused {
```
Does the container have to not exist in runc for restore to work? If so, we should restrict this to ContainerStateConfigured and ContainerStateExited

mheon commented:
Couple small things, looks solid overall. @rhatdan @baude @TomSweeneyRedHat PTAL

TomSweeneyRedHat commented on docs/podman-container-checkpoint.1.md:
```diff
@@ -0,0 +1,30 @@
+
+Keep all temporary log and statistics files created by CRIU during checkpointing. These files
+are not deleted if checkpointing fails fur further debugging. If checkpointing succeeds these
+files are theoretically not needed, but if these files are needed podman can keep the files
```
Prefer "Podman" unless being used as part of a command.

TomSweeneyRedHat commented on docs/podman-container-restore.1.md:
```diff
@@ -0,0 +1,34 @@
+
+Keep all temporary log and statistics files created by CRIU during
+checkpointing as well as restoring. These files are not deleted if restoring
+fails fur further debugging. If restoring succeeds these files are
```
fur to for here too.

TomSweeneyRedHat commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	// Let's try to stat() CRIU's inventory file. If it does not exist, it makes
+	// no sense to try a restore. This is a minimal check if a checkpoint exist.
+	if _, err := os.Stat(filepath.Join(c.CheckpointPath(), "inventory.img")); os.IsNotExist(err) {
+		return errors.Wrapf(err, "It seems there does not exist a complete checkpoint, cannot restore")
```
Suggest: "A complete checkpoint for this container cannot be found."

adrianreber commented on libpod/container_internal_linux.go:
```diff
@@ -307,3 +313,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	// Save network.status. This is needed to restore the container with
+	// the same IP. Currently limited to one IP address in a container
+	// with one interface.
+	formatJSON, err := json.MarshalIndent(c.state.NetworkStatus, "", "	")
```
CNI fails with:

Error adding network: failed to allocate for range 0: requested IP address 10.22.1.236 is not available in range set 10.22.0.1-10.22.255.254

So it fails to restore the container, so yes it is handled, kind of...

adrianreber commented:
@mheon I changed all the points from your review. Thanks for the feedback.

@TomSweeneyRedHat I also fixed everything you found, thanks. Concerning a blog post, I am happy to write an article once this is ready. Would podman.io be the right place for this?

All changes have been pushed, let's see if the CI still likes the current pull request.

rhatdan commented:
LGTM

rhatdan commented:
@adrianreber Sorry but this PR needs a rebase.

adrianreber commented:
Rebased and pushed

adrianreber commented:
Any further comments or changes necessary? Ready to be merged?

mheon commented:
I think we're good to merge

mheon commented:
/lgtm

mheon commented:
/approve

TomSweeneyRedHat commented:
LGTM, @adrianreber TYVM for getting this all together and for addressing all the comments.

adrianreber commented:
Thanks a lot for keeping on reviewing my changes and please let me now if I broke something. Thanks everyone! I have already the next patches prepared for checkpoint/restore in podman and will soon open the next PR(s).

warmchang commented:
Very interesting and useful features. 👍
